### ISSUE TITLE
Property 'validate' in type 'Validator' is not assignable to the same property in base type 'Model<any, ModelSetOptions

### ISSUE THREAD

**Issue by Pavan90** [clientIO/joint#1417]
Hi, 

I am using rappid js version 3.2 from months which works smooth as it should be but since from yesterday my build is failing and facing the below error 

`ERROR in node_modules/@clientio/rappid/rappid.d.ts:6046:9 - error TS2416: Property 'validate' in type 'Validator' is not assignable to the same property in base type 'Model<any, ModelSetOptions, {}>'.
  Type '(actions: string, ...callback: ((err: Error, command: any, next: any) => boolean)[]) => Validator' is not assignable to type '(attributes: Partial<any>, options?: any) => any'.
    Types of parameters 'actions' and 'attributes' are incompatible.
      Type 'Partial<any>' is not assignable to type 'string'.

6046         validate(actions: string, ...callback: Array<(err: Error, command: any, next: any) => boolean>): Validator;`

Its coming from rappid.d.ts  which is a file inside node_modules. Not sure anything got changed from the rappidJS end. 

We use Angular 10. 

Please let me know how to solve this one or point me to the Issues where it was a known issue with any suggested fixes. Thanks.

**Comment by kumilingus** [comment]
Perhaps one of the typings dependencies has changed? What version of @types for `Backbone` do you have in `package.json`?

**Comment by Pavan90** [comment]
Thanks for the response. We are using     "backbone": "^1.4.0"

**Comment by kumilingus** [comment]
I meant `@types/backbone` actually. 
I can see there was a release yesterday - https://www.npmjs.com/package/@types/backbone).
Could you try to downgrade this dependency?

**Comment by Pavan90** [comment]
oh sorry. Here is the one we looking for     "@types/backbone": "^1.4.4". I will downgrade it and give a shot and let you know.

**Comment by Pavan90** [comment]
I have used @types/backbone": "1.4.5" and the type error is gone now. I guess they have to fix their types issue on their latest package. I will let them know about this issue with their latest package version on their repo. Thanks for your awesome help. Please close this issue.

**Comment by Pavan90** [comment]
I have checked the @types/backbone index.d.ts file where they are following types for validate method  from the latest package release.  From the below line of code, before, the first parameter of validate method is a type any but from latest release, the type changed from any to "Partial".  

**`validate(attributes: Partial<T>, options?: any): any;**
`
but in the rappid.d.ts file I can see that validate first parameter is a type of string which causes the issue now. Which worked fine before because of the type 'any'. 

 **validate(actions: string, ...callback: Array<(err: Error, command: any, next: any) => boolean>): Validator;**

Either we have to make the actions parameter type as 'any' or Partial<any> whatever the type works better.

**Comment by kumilingus** [comment]
I am trying to reproduce the error. How do you call `validate()` method in your app?

I don't get why typescript complains about this. The `Validator` inherits from `Backbone.Model` class and overrides one of its method (`validate()`) completely (changing the method's signature).

**Comment by Pavan90** [comment]
I did scanned the entire repo and no where I or my co devs have used this method in their components. My docker image will get created by doing fresh npm install everytime and caught that error. Later I deleted node_modules and did fresh install and did ng build /ng serve and started seeing the error from rappid.d.ts file. 


Basically this is the class we need to fix 

`    class Validator extends Backbone.Model {

        constructor(opt: Validator.Options);

        validate(actions: string, ...callback: Array<(err: Error, command: any, next: any) => boolean>): any;
    }`

You are right, Validator extends and overrides itself but its method validate(actions: string ) still takes string rather than type Partial <T> or Partial<any> or any. We can fix this by updating the type of parameters of validate method. In the previous types/backbone version, they were using any but now they made its a strict type which is partial

**Comment by kumilingus** [comment]
Hmm, let me ask you this way. Should this throw an error? I must be missing something.
```ts
class A {
  test(param: number): void;
}

class B extends A {
  test(param: string): void;
}
```

**Comment by Pavan90** [comment]
look at here https://www.typescriptlang.org/play?#code/FAYwNghgzlAECCsDexawC4FMroBQAcIAnCAWwC5YA7AV1ICNMiBKSgNwHsBLAE2QF8A3MH7BQkGLABCsTAA8sVHnEQo0WHAWJlKOIlyoBzVrE68Bw0eOhwAwgB4AKrAC8AgHzJU67HkIkKWAAFYnQuCDAndxMzPiQhETFwG1gAEVkFTCU7LzQMXy0AyloGJhjuOITRIA

**Comment by kumilingus** [comment]
Thanks :) I'll fix it it in Rappid for the upcoming release.

**Comment by Pavan90** [comment]
You’re welcome 🙌

**Comment by kumilingus** [comment]
Fixed in Rappid `v3.3`.

**Comment by Pavan90** [comment]
Sweet! Thanks
